### ISSUE TITLE
Link to a revision page before a rename on the history page is super slow (Mercurial).

### ISSUE THREAD

**Issue by wangweij** [OpenGrok/OpenGrok#1765]
It's very nice that the history page of a file contains the old history before the file is renamed. However, The revision link still uses the new name and navigating to the page is super slow, and sometimes just fails. It will be nice if the link generated is already using the old name.

**Comment by vladak** [comment]
I gather that you're suggesting to cache the original name. It's true that the way it is implemented can be called hackish (or opportunistic). The `getHistoryGet()` method in SCMs that support this (e.g. Mercurial) try to fetch the contents of the file in given revision using the current name first and if that fails, attempt to find the original name in given revision is made and if this succeeds, the file contents are fetched. What probably takes bulk of the time is the display of verbose log for given file that requires to run the SCM command which is usually I/O intensive, thus inducing the low performance. Both command invocations can probably be skipped if `HistoryEntry` for renamed files contained name of the file in each revision.

**Comment by wangweij** [comment]
I tried both methods on https://stackoverflow.com/questions/7948113/how-can-you-see-the-previous-name-of-a-renamed-file-in-mercurial and they are very slow. Here I am working on a big repo that moved all files into a new place in a changeset. That is the bottleneck.

**Comment by vladak** [comment]
Yep, once Mercurial needs to display list of modified files (result of -v option or `file_copies` template), it needs to dig deeper and this incurs significant I/O, making the whole operation slower than simple changeset display. The bottom line is that in order to determine the original name of renamed file in given changeset, one has to traverse all the changesets from the current one to the one in question and follow the renames.

**Comment by vladak** [comment]
Marking for 1.1 however this will likely induce historyindex change (`HistoryEntry` augmentation as explained above) so will address it late in the release cycle in order to make it easier to use the release candidates (without complete reindex).
